### PR TITLE
Added module for baking EC2 AMIs

### PR DESCRIPTION
**Pull request by eduffield222**
I added a module for baking an AMI from a running EC2 instance and deregistering/deleting them. Let me know if any other changes need to be made to merge into core. Thanks

### CODE REVIEW

**Comment by mpdehaan** [comment]
@lwade @jarv would you like to help take a look at this one?

Thanks!

File: library/cloud/ec2_ami [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            ec2_access_key = dict(aliases=['EC2_ACCESS_KEY']),
+            instance_id = dict(),
+            image_id = dict(),
+            delete_snapshot = dict(),
+            name = dict(),
+            wait = dict(type="bool", default=False),
+            wait_timeout = dict(default=300),
+            description = dict(default=""),
+            no_reboot = dict(default=True, type="bool"),
+            state = dict(default='present')
```
Comment (jarv) [inline library/cloud/ec2_ami]: is `region` missing from the argument_spec?

File: library/cloud/ec2_ami [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            module.fail_json(msg = str(e))
+
+    if module.params.get('state') == 'absent':
+        if not module.params.get('image_id'):
+            module.fail_json(msg='image_id needs to be an ami image to registered/delete')
+
+        deregister_image(module, ec2)
+
+    elif module.params.get('state') == 'present':
+        # Changed is always set to true when provisioning new instances
```
Comment (jarv) [inline library/cloud/ec2_ami]: comment is a little misleading I think, should it be "provisioning new AMIs?"

File: library/cloud/ec2_ami [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+try:
+    import boto.ec2
+    from boto.exception import EC2ResponseError
+except ImportError:
+    print "failed=True msg='boto required for this module'"
+    sys.exit(1)
+
+def create_image(module, ec2):
+    """
+    Creates new instances
```
Comment (jarv) [inline library/cloud/ec2_ami]: I would change this to "Create new image"

File: library/cloud/ec2_ami [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    """
+
+    image_id = module.params.get('image_id')
+    delete_snapshot = module.params.get('delete_snapshot')
+    wait = module.params.get('wait')
+    wait_timeout = int(module.params.get('wait_timeout'))
+
+    img = ec2.get_image(image_id)
+    if img == None:
+        module.fail_json(msg = "Image %s does not exist" % image_id)
```
Comment (jarv) [inline library/cloud/ec2_ami]: If the image doesn't exist we might want to exit here with changed=False

File: library/cloud/ec2_ami [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    delete_snapshot: False
+    state: absent
+
+'''
+import sys
+import time
+
+try:
+    import boto.ec2
+    from boto.exception import EC2ResponseError
```
Comment (jarv) [inline library/cloud/ec2_ami]: I don't see this exception used anywhere, is the import necessary?

File: library/cloud/ec2_ami [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+    if module.params.get('state') == 'absent':
+        if not module.params.get('image_id'):
+            module.fail_json(msg='image_id needs to be an ami image to registered/delete')
+
+        deregister_image(module, ec2)
+
+    elif module.params.get('state') == 'present':
+        # Changed is always set to true when provisioning new instances
+        changed = True
```
Comment (jarv) [inline library/cloud/ec2_ami]: the variable `changed` doesn't look like it's used

File: library/cloud/ec2_ami [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+                  'no_reboot': no_reboot}
+
+        image_id = ec2.create_image(**params)
+    except boto.exception.BotoServerError, e:
+        module.fail_json(msg = "%s: %s" % (e.error_code, e.error_message))
+
+    # wait here until the image is gone
+    img = ec2.get_image(image_id)
+    wait_timeout = time.time() + wait_timeout
+    while wait and wait_timeout > time.time() and (img == None or img.state != 'available'):
```
Comment (jarv) [inline library/cloud/ec2_ami]: Better to use "is" when comparing to None

**Comment by jarv** [comment]
some nitpicks, looks good to me otherwise.  We will definitely find this useful, thanks for writing it!

**Comment by eduffield222** [comment]
Great thanks! I can fix this all in the morning

**Comment by lwade** [comment]
Good comments by @jarv . Nice addition.  Definately need that region support though :)

**Comment by eduffield222** [comment]
Any word on this?

File: library/cloud/ec2_ami [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+            ec2_access_key = dict(aliases=['EC2_ACCESS_KEY']),
+            instance_id = dict(),
+            image_id = dict(),
+            delete_snapshot = dict(),
+            name = dict(),
+            wait = dict(type="bool", default=False),
+            wait_timeout = dict(default=900),
+            description = dict(default=""),
+            no_reboot = dict(default=True, type="bool"),
+            state = dict(default='present')
```
Comment (jarv) [inline library/cloud/ec2_ami]: I think we need to add region to the argument_spec

**Comment by eduffield222** [comment]
Added region to the specs, let me know if there is anything else that needs to be done .Thanks!

**Comment by pas256** [comment]
This looks good to me too.

I am not sure how many people make AMIs public, but it may be worth adding a comment about ensuring the security of it. In general, making a public AMI off a running EC2 instance is a bad idea because it leaves a lot of information around. See http://alestic.com/2011/06/ec2-ami-security

Thanks Evan.

**Comment by eduffield222** [comment]
Boto by default makes the AMI you create private -- I just did a test to double check. However, I'll put a note in the source about it.

**Comment by mpdehaan** [comment]
merged, thank you!  very cool.

**Pull request merged**